### ISSUE TITLE
Project Meteor: separate add-on update, download, and install functions from the user interface

### ISSUE THREAD

**Issue by josephsl** [josephsl/addonUpdater#11]
Hi,

I expect this will take the entire northern summer:

## Background
Add-on Updater's update check, download, and install routines are mostly based on NVDA's own update check mechanism. This is evidenced by the fact that add-on update downloader class inherits from NVDA Core's downloader class. However, this means people wishing to test download and install routines cannot do so unless they invoke add-on update check routine and if add-on updates are indeed available.

## Project Meteor overview
The overall goal of Project Meteor is to separate add-on update check, download, and install routines from the accompanying graphical user interface. As of time of this writing, update check routine and the progress dialog for it are separated, allowing people to test the update check mechanism without having to select "Check for add-on updates" entry from NVDA tools menu. This is possible as the update check routine returns a dictionary of update records, and the only thing the progress dialog does is call the actual routine through a thread. This should be extended to add-on update download and install routines so they can be checked independent of the GUI.

### Before and after
Currently, Add-on Updater does the following when installing add-on updates:

1. NVDA retrieves two JSON files: add-on files repository (get.php) containig download links for add-ons, and a compatibility database (NVDA add-ons repo on GitHub) that records add-on compatibility information such as minimum NVDA version.
2. The retrieval routine is call from addonHandlerEx.checkForAddonUpdates function, which in turn uses threads to retrieve the JSON files and evaluate update information such as version and compatibility checks. Threads are used to prevent main thread freezing which can cause NVDA to appear to do nothing.
3. If the user chooses to check for add-on updates from the GUI, steps 1 and 2 are performed while a progress dialog is active. The overall update check GUI coordinator is a separate function that will use threads to perform steps 1 and 2 and destroy the dialog once the update check is complete.
4. If updates are available, add-on updates dialog launches or a toast is shown depending on user settings.
5. After the user chooses add-ons to be updated, a generator is executed that will create an instance of add-on update downloader class for each add-on update entry.
6. Add-on update downloader first downloads the add-on installer package, which borrows heavily from NVDA Core's update download routine.
7. If download succeeds, the newly downloaded add-on is installed via add-on downloader class.
8. If at least one add-on update succeeds, NVDA will prompt the user to restart it so changes can take effect.

With Project Meteor:
1. Follow steps 1 through 4 from above.
2. Use a different function to download add-ons inside a thread, with the main thread using a progress dialog to keep an eye on download progress.
3. The downloader routine will be based on the download routine from add-on downloader class with a difference being that temp path will be returned or added to an add-on update record (depending on implementation).
4. Once an add-on is downloaded, either download other add-ons or move onto update install routine.
5. Similar to update check routine, a progress dialog will be used to keep an eye on a separate add-on install function. The update install function will take add-on name and temp path as arguments and can either launch error dialogs or return status flags (depends on implenentation).
6. After all add-ons are updated, if at least one add-on was updated successfully, ask users to restart NVDA.

### Secondary goals
While the primary goal of Project Meteor is to separate download and installation routines from the GUI, secondary goals include:

* Easier testing: instead of using GUI to test add-on update routine, testers can use Python Console to test download and install routines separately.
* Maintenance: instead of having to edit functions that will perform update operation and GUI presentation at the same time, using separate functions (and classes if needed) allows implementation to change without affecting other components. For example, if the update routine somehow changes from NVDA Core, only the install routine from this add-on need to change without affecting the GUI. In turn, this allows testers to test edits without affecting overall functionality of the add-on - for example, testing an implementation which checks SHA values for ad-on packages and returning an appropriate error.
* Add-on update automation: while not an immediate goal, defining routines for download and install without relying on GUI opens up a possibility to automate add-on updates in the future by just executing update check/download/install routines without invoking GUI.
* Downloading all add-ons and then installing them: currently add-ons are downloaded and then instlaled one at a time. If Project Meteor succeeds, allow downloading all add-on updates at once and then launch the update routine.

## Project schedule
This is a tentative schedule for Project Meteor implementation:

1. Major bug fixes (milestone 1): in the course of researching this issue, several major bugs were found in current download and install routines, including download progress dialog not centered on screen. These bugs should be identified and fixed before Project Meteor can actually commence, with the milestone target being Add-on Updater 22.07.
2. Create a skeleton (milestone 2): isolate Project Meteor from the rest of the add-on through a new module named addonUpdateProc (add-on update processes and procedures). This module will house skeletons (function/class prototypes) of update check/download/install routines which does nothing. Milestone target is about a week after Add-on Updater 22.07 release.
3. Fill the skeleton (milestone 3): fil the just defined skeletons with routines coming from extended add-on handler and GUI modules which houses update check and download/install routines, respectively. This should be done as the last step in milestone 2.
4. Transform routines (milestone 4): for each routine, identify and separate core routines from GUI code. Milestone target is a few days after milestone 3.
5. Add-on update record reorganization (optional milestone 4.1): define a class that houses add-on update record, currently housed inside dictionaries. The record instance will include add-on name, summary, current version, update version, download URL, temp path, hash if any, compatibility information, and other routines if defined in the future. For backward compatibility, define a dictionary method that will return add-on update record information as a dictionary.
6. Test update record classes (optional milestone 4.2): test the new add-on update record classes with the just defined update check/download/install routines. Milestones 4.1 and 4.2 should be done after milestone 4 is complete.
7. Prepare for broad testing (milestone 5): use a combination of conditional code, variables, and command-line switches to let users test the old and the new routines for a while.
8. Interaction with the rest of the add-on (milestone 6): after gathering data from test builds, allow other parts of the add-on to use the new routines in appropriate places. Milestone target is a few days after milestone 5.
9. Integration testing (milestone 7): release test builds to test integration work. From this milestone forward, old routines will be gone.
10. Refinements: refine Project Meteor based on feedback from the community.
11. Stable channel release: release Project Meteor Add-on Updater to stable channel users, tentatively in August if the above schedule goes as planned.
12. Continued maintenance: refine Project Meteor and secondary oals based on user feedback.

## Alternatives
At least fix major bugs and no more since the add-on code is complex at this time.

Thanks.

**Comment by josephsl** [comment]
Hi,

Update as of July 5th: milestone 5 is in progress and project core is done (download and install routines are now truly separate from GUI). Project Meteor core will become part of main branch as early as July 8th with user interface to follow.

Also, another secondary goal has emerged: different add-on update sources. This requires refactoring update check mechanism so it can fetch update metadata from different places (at the moment Add-on Updater is optimized to fetch things from community add-ons website). The resulting work, tentatively named "protocols", will allow Add-on Updater to update add-ons from different sources (with caveats and warnings).
Thanks.

**Comment by josephsl** [comment]
Hi,
Update as of July 6th: Project Meteor core is complete. Next stpes include integration and the corresponding user interface.

Thanks.

**Comment by josephsl** [comment]
Hi,

Resolved in version 22.09.
